internal/engine/internal/oobackend/client.go

```diff
@@ -0,0 +1,112 @@
+package oobackend
+
+import (
+	"bytes"
+	"errors"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/ooni/probe-cli/v3/internal/engine/internal/multierror"
+)
+
+// TODO(bassosimone): apply similar changes to sessionresolver. In
+// particular, let's avoid writing state in sessionresolver.Resolver
+// by making the KVStore mandatory, which is a small change.
+
+// Client is a client to speak to the OONI backend. In its
+// default configuration (i.e., with no proxies), this client
+// uses the best strategy to speak to the backend. Make sure
+// to fill the mandatory fields before using a Client.
+type Client struct {
+	// KVStore is the mandatory KVStore from which we read
+	// the state and in which we write the state.
+	KVStore KVStore
+
+	// HTTPClientWithProxy is an optional HTTP client that is
+	// configured to always use a proxy. When this client is
+	// present, we always and only use it, and we do not update
+	// our strategy dynamically. In fact, when this client is
+	// present, it means the user has requested the probe to
+	// use a specific proxy explicitly.
+	HTTPClientWithProxy HTTPClient
```

Comment:
I am wondering whether a better design would be to have a boolean indicating we're using a proxy. Another possible design here could be that we don't use this client in case we are using a proxy, but we use the real HTTP client directly.